test/prober/await.go

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright 2020 The Knative Authors
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package prober
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"go.opencensus.io/trace"
+	"go.uber.org/zap"
+	appsv1 "k8s.io/api/apps/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"knative.dev/eventing/test/lib"
+	"knative.dev/eventing/test/lib/duck"
+	"knative.dev/eventing/test/lib/resources"
+	"sync"
+	"time"
+)
+
+func (p *prober) waitForKServiceReady(name, namespace string) error {
+	meta := resources.NewMetaResource(name, namespace, &servingType)
+	return duck.WaitForResourceReady(p.client.Dynamic, meta)
+}
+
+func (p *prober) waitForKServiceScale(name, namespace string, satisfyScale func(*int32) bool) error {
+	metricName := fmt.Sprintf("waitForKServiceScale/%s/%s", namespace, name)
+	_, span := trace.StartSpan(context.Background(), metricName)
+	defer span.End()
+
+	return wait.PollImmediate(duck.Interval, duck.Timeout, func() (bool, error) {
+		serving := p.client.Dynamic.Resource(servicesCR).Namespace(namespace)
+		unstruct, err := serving.Get(name, metav1.GetOptions{})
+		return p.isScaledTo(satisfyScale, unstruct, namespace, err)
+	})
+}
+
+func (p *prober) isScaledTo(satisfyScale func(*int32) bool, un *unstructured.Unstructured, namespace string, err error) (bool, error) {
```

Comment:
This function is a bit hard to understand and pretty brittle I think...

Is it only being used to check `ScaleToZero` and `ScaleFromZero`? If yes, https://github.com/knative/serving/blob/637dc521323f908362c290da66bf9ebd3c498b64/test/performance/performance.go#L93-L133 might help, though we might want them to be moved to `pkg/test` before using.

Reply:
Yes. It's just for `scaleToZero`. I write that down using Dynamic client as you asked here: https://github.com/knative/eventing/issues/2271#issuecomment-573502475

That's the reason why I'm not using that as it is in `knative/serving`. I didn't vendored `knative/serving`.

Reply:
If they are doing the same thing, can we copy the two functions to this PR and use it here?
We can have a comment saying they need to be moved to a common repo, which we will do after this PR is merged.

Reply:
I can't use it without vendoring `knative/serving`.

I use Dynamic client to fetch and traverse ksvc and to get its deployment. After I got deployment I do the same thing as does function you are pointing - that's getting number of ready replicas from corev1 Deployment API.

Reply:
I refactored those methods, and moved them to lib package